src/Razor/src/Microsoft.VisualStudioCode.RazorExtension/Services/MappingService.cs

```diff
@@ -0,0 +1,99 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.ExternalAccess.Razor;
+using Microsoft.CodeAnalysis.Razor;
+using Microsoft.CodeAnalysis.Razor.Protocol;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.VisualStudioCode.RazorExtension.Services;
+
+internal class MappingService(IRazorClientLanguageServerManager razorClientLanguageServerManager) : IRazorMappingService
+{
+    private const string RazorMapSpansEndpoint = "razor/mapSpans";
+    private const string RazorMapTextChangesEndpoint = "razor/mapTextChanges";
+
+    private readonly IRazorClientLanguageServerManager _razorClientLanguageServerManager = razorClientLanguageServerManager;
+
+    public async Task<ImmutableArray<RazorMappedSpanResult>> MapSpansAsync(Document document, IEnumerable<TextSpan> spans, CancellationToken cancellationToken)
+    {
+        if (string.IsNullOrWhiteSpace(document.FilePath))
+        {
+            return [];
+        }
+
+        var sourceText = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+
+        var mapParams = new RazorMapSpansParams()
+        {
+            CSharpDocument = new()
+            {
+                Uri = new(document.FilePath)
+            },
+            Ranges = [.. spans.Select(sourceText.GetRange)]
+        };
+
+        var response = await _razorClientLanguageServerManager.SendRequestAsync<RazorMapSpansParams, RazorMapSpansResponse?>(
+            RazorMapSpansEndpoint,
+            mapParams,
+            cancellationToken).ConfigureAwait(false);
+
+        if (response is not { Spans.Length: > 0, Ranges.Length: > 0 })
+        {
+            return [];
+        }
+
+        Debug.Assert(response.Spans.Length == spans.Count(), "The number of mapped spans should match the number of input spans.");
+        Debug.Assert(response.Ranges.Length == spans.Count(), "The number of mapped ranges should match the number of input spans.");
+
+        var builder = ImmutableArray.CreateBuilder<RazorMappedSpanResult>(response.Spans.Length);
```

Comment:
Consider using a `PooledArrayBuilder<RazorMappedSpanResult>` to avoid this allocation.

```suggestion
        using var builder = new PooledArrayBuilder<RazorMappedSpanResult>(response.Spans.Length);
```